### ISSUE TITLE
Horovod hangs on Keras TensorBoard Histograms

### ISSUE THREAD

**Issue by nicolas-ivanov** [horovod/horovod#938]
**Environment:**

Horovod docker container 0.16.0-tf1.12.0-torch1.0.0-mxnet1.4.0-py3.5

1. Framework: TensorFlow, Keras
2. Framework version: tensorboard==1.12.2, tensorflow-gpu==1.12.0, Keras==2.2.4
3. Horovod version: horovod==0.16.0
4. MPI version: (OpenRTE) 3.1.2
5. CUDA version: V9.0.176
6. NCCL version: 2.4.2
7. Python version: 3.5.2
8. OS and version: Ubuntu 16.04.5 LTS

**Checklist:**
- [x] Did you search issues to find if somebody asked this question before? – Yes
- [x] If your question is about hang, did you read [this doc](https://github.com/horovod/horovod/blob/master/docs/running.md)? – Yes
- [ ] If your question is about docker, did you read [this doc](https://github.com/horovod/horovod/blob/master/docs/docker.md)? – No

**Bug report:**

It's impossible to log gradients and weights histograms using Keras TensorBoards callback while training the model on multiple GPUs.

Gist to reproduce the bug: https://gist.github.com/nicolas-ivanov/bcba768c27ee0c2735b0497cd060c994

Command to run:
```
CUDA_VISIBLE_DEVICES='0,1 mpirun -np 2 -H localhost:2 python mnist_tb_with_horovod.py
```

Output after some time:
```
...
Epoch 2/24
[2019-03-19 20:46:02.219911: W horovod/common/operations.cc:1669] One or more tensors were submitted to be reduced, gathered or broadcasted by subset of ranks and are waiting for remainder of ranks for more than 60 seconds. This may indicate that different ranks are trying to submit different tensors or that only subset of ranks is submitting tensors, which will cause deadlock. 
Stalled ops:DistributedAdadelta_Allreduce_3/HorovodAllreduce_gradients_3_conv2d_2_BiasAdd_grad_BiasAddGrad_0 [missing ranks: 1]
[2019-03-19 20:46:02.220006: W horovod/common/operations.cc:1669] DistributedAdadelta_Allreduce_2/HorovodAllreduce_gradients_2_conv2d_2_convolution_grad_Conv2DBackpropFilter_0 [missing ranks: 1]
[2019-03-19 20:46:02.220029: W horovod/common/operations.cc:1669] DistributedAdadelta_Allreduce_5/HorovodAllreduce_gradients_5_dense_1_BiasAdd_grad_BiasAddGrad_0 [missing ranks: 1]
[2019-03-19 20:46:02.220049: W horovod/common/operations.cc:1669] DistributedAdadelta_Allreduce_1/HorovodAllreduce_gradients_1_conv2d_1_BiasAdd_grad_BiasAddGrad_0 [missing ranks: 1]
...
```

Running the provided code on 1 GPU or with `histogram_freq=0` doesn't hang up the training process.

**Comment by alsrgv** [comment]
@nicolas-ivanov, can you check if discussion in #297 helps?

**Comment by nicolas-ivanov** [comment]
@alsrgv, thanks for your reply. Yes, I saw the mentioned thread, but I believe it is irrelevant to my case since I do provide validation_data batch (see the [line](https://gist.github.com/nicolas-ivanov/bcba768c27ee0c2735b0497cd060c994#file-mnist_tb_with_horovod-py-L128)).

**Comment by alsrgv** [comment]
@nicolas-ivanov, I've debugged this issue a bit and I think I found the root cause.

Since you have `write_grad=1` set, the TensorBoard callback will query optimizer gradients.  Optimizer gradients are computed using allreduce.  However, since you're only writing TensorBoard logs on the first rank, it waits forever for the other ranks to run allreduce.

If you don't need `write_grads`, you can disable it.  If you do want to see gradients, you need to run TensorBoard callback on all processes - just make sure they use different directories to avoid conflicts.

**Comment by nicolas-ivanov** [comment]
@alsrgv thank you for digging into this issue! Yes, I do need to write gradients to TensorBoard. I'll check your suggestion with running TensorBoard callback on each worker, however, do you think there can be a more elegant solution for this? I thought, Horovod could handle such cases automatically.

**Comment by alsrgv** [comment]
@nicolas-ivanov, unfortunately, running TensorBoard callback on all workers is the most elegant solution so far.  We do have plans to enable `hvd.join()` operation (see #832), which could help here as well.  I'll mark this issue as an enhancement for future.

**Comment by nicolas-ivanov** [comment]
@alsrgv Got you, thanks once again for the explanation!
